### DISCUSSION TITLE
Enhanced abbreviations

### DISCUSSION THREAD

**Discussion by ridiculousfish** [fish-shell/fish-shell#8854] [Ideas]
Here's some ideas around enhanced abbreviations. I intend to implement these if others agree. Let me know what you think!

Recall that [abbreviations](https://fishshell.com/docs/3.4/cmds/abbr.html) have a name and a replacement. If the user types an abbreviation in command position, it gets replaced with a value. `abbr gco 'git checkout'` is an example.

I would like to extend this feature to give users a high degree of customization, similar to key bindings. Here's my proposed signature:

    abbr --add NAME [-p|--position command|anywhere] [-o|--expands-on entry|execute] [-r|--regex PATTERN] [-f|--function] REPLACEMENT

Here's what these new options do:

### --position command | anywhere

Today abbreviations only expand in "command position." If you pass `--position anywhere` they expand anywhere. The default remains `command`. See #5003.

### --expands-on entry | execute

Today an abbreviation expands after the user types a space, or enter. If you pass `--expands-on execute` it expands only on enter. History receives the command *before* expansion, allowing abbreviations to work better with history and autosuggestions.

The default remains `entry`.

### --regex PATTERN

Today, an abbreviation replaces its name with its value. With `--regex` the user can supply an arbitrary regex to describe the token to replace. The abbreviation still has a name so it can be identified, but it doesn't contribute to matching.

Initially only single tokens can be matched.  We'll use PCRE2 of course.

If multiple abbreviations have a matching regexp, the last one takes precedence. This means the order of abbreviations may matter.

### --function

Lastly, the `--function` flag indicates that REPLACEMENT should be interpreted as a fish function name. To generate the replacement, fish invokes the function, passing it the token which matched. If the function succeeds (returns 0) its output is used to replace the token. If it fails, no replacement occurs.

## Examples

Some examples of stuff you can do. Not saying any of these are good ideas, except the first I think we should ship as an example.

We can easily support `!!` (see #288)

```
function bangbang
     echo $history[1]
end
abbr !! --location anywhere --expands-on  execute --function bangbang
```

You can go a little nuts and implement more of history expansion:

```
function histexpand
    # Receives a token like !-5.
    # Strip off the !- and substitute in the index.
    echo $history[(string sub -s 3 -- $argv)]
end
abbr histexpand --location anywhere --expands-on execute  --regex !-\\d+ --function histexpand
```

You can also use it as a sort of superpowered key binding. This enables `<E...>` syntax to insert emojis by name:

```
> function emojize
    set name (string match --groups-only -r <E (.+)> -- $argv)
    somehow_get_emoji_by_name
end
> abbr emojize --location anywhere --expands-on  entry  --regex <E .+> --function emojize
> <E GREEN HEART> # replaced with corresponding emoji
```

## Other notes

Abbreviations will no longer be managed by universal variables; instead they will be managed internally like key bindings. This is because abbreviations now have a lot more metadata which is unwieldy to attach to just variables; it's also consistent with our desire to move away from uvars.

The existing behavior where `abbr` sets a universal variable will no longer work. Instead users will have to add abbreviations to `config.fish` or elsewhere. The `-U` and `-g` flags will be ignored. We will continue to import existing universal-variable backed abbreviations, so we don't break existing configs.

Abbreviations remain an interactive feature only. They do not work in scripts.

Thank you for any thoughts!

**Comment by krobelus** [comment]
Looks really solid. I have some thoughts but no objections.

I understand (regex) matching is done on the unexpanded tokens (so a literal `"!!"` is not matched), and the entire token is replaced.
Having users deal with unexpanded tokens always feels dirty but that's unlikely to cause problems here.

I wonder if/how we want to cover abbreviations for subcommands (https://github.com/fish-shell/fish-shell/issues/6283).
Let's say we want `g co` -> `git checkout` and `s co` -> `string collect`. Since abbreviations are only matched against the current token, we can't have both meanings of `co`. One possible solution is to let abbreviations run arbitrary code instead of `--regex`. Barring a good use case, this doesn't seem worth it.

Vim expands abbreviations at word boundaries. We expand on space.
The Vim approach has some advantages, for example given `abbr hp --location anywhere /path/to/my/happy-place` I could type `cd hp/abc` and it expands as soon as I type the `/`. This sounds like something we should try to do once we have global abbreviations.

Couple more examples:

	abbr --location=anywhere '"$@"' $argv
	abbr --location=anywhere '"$*"' (string join ' ' -- $argv)
	abbr --location=anywhere '$?' $status
	abbr --location=anywhere '$!' $last_pid
	abbr --location=anywhere '$#' (count $argv)
	abbr --location=anywhere '$$' $fish_pid
	function fix-braced-variable
		printf '{$%s}' (string match --groups-only -r -- '\$\{(\w+)\}' $argv)
	end
	abbr --location=anywhere '${VAR}' --regex '\$\{\w+\}' --function fix-braced-variable

Most importantly they have low risk of false positives, because the inputs are invalid syntax.

Some of these would benefit from abbreviation-expansion at word boundaries.
However, it's nontrivial to detect whether the input is escaped/quoted, so I'd grudgingly stick to the simple, hardcoded patterns above taht only match whole tokens instead of trying to support like `echo "error: $?"`.

If we could do it without a breaking change, I think `--location=anywhere` should be the default.

**Reply by faho** [reply]
>If we could do it without a breaking change, I think --location=anywhere should be the default.

I disagree.

I see common uses of abbrs to be something like "usc -> systemctl --user" or "g -> git". I certainly don't want "g" to be changed to "git" everywhere.

**Comment by lacygoill** [comment]
Thank you very much for working on this.

   > Thank you for any thoughts!

FWIW, I would like to discuss 2 other features.

   > Today an abbreviation expands after the user types a space, or enter. If you pass --expands-on execute it expands only on enter. History receives the command before expansion, allowing abbreviations to work better with history and autosuggestions.

I would be interested in `--expand-on space` to limit the expansion after a space.  No expansion after pressing Enter.  Rationale:

   - I don't like the shell running something different than what I can read right before pressing Enter.
   - Expanding after `Enter` can lead to confusing results.

Regarding the 2nd bullet point, consider this:

    $ mkdir -p /tmp/empty
    $ XDG_CONFIG_HOME=/tmp/empty fish
    $ abbr ls 'FOO=bar ls'

    $ ls

`ls` is expanded into this:

    FOO=bar ls

So far, so good.

Now, suppose you made a typo, and you want to replace `bar` with `baz`.  You press `CTRL-P` to recall the last command, and edit the command as necessary:

    FOO=baz ls

Finally, you press Enter.
Expected: fish runs `FOO=baz ls`.
Actual: fish runs `FOO=baz FOO=bar ls`.

I know it's working as intended, but I still find this confusing.  I would like the ability to limit the expansion to after a space only.

---

It might also be useful to let us specify the cursor position after the expansion.

For example:

    $ abbr --add --global gc 'git clone --depth=1   ; cdd'

With `cdd` being a custom function to cd into the most recent subdirectory in the current working directory:

    function cdd
        # `%Z` will be replaced with the time of last status change (in seconds since Epoch).
        # `%n` will be replaced with the file name.
        set -f most_recent_dir $(
            fd --type=d --max-depth=1 --exec stat --format '%Z %n' {} \
          | sort --key=1,1 --numeric-sort --reverse --ignore-leading-blanks \
          | head --lines=1 \
          | awk '{print $2}'
        )
        cd $most_recent_dir
    end

Suppose you want to clone the fish project.  You type `gc` then space to trigger the expansion, and you get this command-line:

    $ git clone --depth=1   ; cdd |
                                  ^
                                  cursor position

Instead, you would like this command-line:

    $ git clone --depth=1 | ; cdd
                          ^
                          cursor position

So that you can paste the url of the fish repo without having to move the cursor first.

This could be achieved with a `--cursor` option, which would expect an integer as argument specifying the position of the cursor in bytes, after the expansion:

                          v------v
    $ abbr --add --global --cursor=20 gc 'git clone --depth=1   ; cdd'
                                   ^^
                                   output of:  string length 'git clone --depth=1 '

Even better would be to let us write a special token inside the expansion to express where we want the cursor to be:

    $ abbr --add --global --cursor gc 'git clone --depth=1 <cursor> ; cdd'
                          ^------^                         ^------^

**Reply by typecasto** [reply]
The cursor position being an integer would be problematic if the abbreviation is executing a function, so I'm more in support of the special token.  Maybe have a --no-move-cursor option in case you need to write a literal \<cursor> for some reason.

**Comment by halostatue** [comment]
One other thing that would be *nice* from my perspective (as I’m trying to retrain more than a decade of Git use to use new porcelain). I’d *love* to be able to perform more complex substitutions that aren’t just single tokens. That is, I find myself typing `git co -b` or `git checkout -b`, when the modern porcelain is `git switch -c`. I think I could do some tweaking with git aliases, but having the correction show up as I’m typing it would be…better.
